File: 2021/src/days/day5.ts
import { Input } from "../input";

type Line = {
  x1: number;
  y1: number;
  x2: number;
  y2: number;
};

const parse = (input: Input): Line[] =>
  input.asLines().map((l) => {
    const [_, ...ps] = l.match(/(\d+),(\d+) -> (\d+),(\d+)/);
    const [x1, y1, x2, y2] = ps.map((p) => parseInt(p, 10));

    return { x1, y1, x2, y2 };
  });

export const part1 = (input: Input) => {
  const lines = parse(input);
  const grid = new Map<string, number>();

  lines.forEach(({ x1, x2, y1, y2 }) => {
    if (x1 === x2) {
      for (let y = Math.min(y1, y2); y <= Math.max(y1, y2); y++) {
        const coord = `${x1},${y}`;
        grid.set(coord, (grid.get(coord) || 0) + 1);
      }
    }
    if (y1 === y2) {
      for (let x = Math.min(x1, x2); x <= Math.max(x1, x2); x++) {
        const coord = `${x},${y1}`;
        grid.set(coord, (grid.get(coord) || 0) + 1);
      }
    }
  });

  return Array.from(grid.values()).filter((x) => x > 1).length;
};

export const part2 = (input: Input) => {
  const lines = parse(input);

  const grid = new Map<string, number>();

  lines.forEach(({ x1, x2, y1, y2 }) => {
    if (x1 === x2) {
      for (let y = Math.min(y1, y2); y <= Math.max(y1, y2); y++) {
        const coord = `${x1},${y}`;
        grid.set(coord, (grid.get(coord) || 0) + 1);
      }

      return;
    }

    const yStep = Math.sign(y2 - y1) * Math.sign(x2 - x1);
    const yStart = x1 < x2 ? y1 : y2;

    let i = 0;

    for (let x = Math.min(x1, x2); x <= Math.max(x1, x2); x++) {
      const coord = `${x},${yStart + yStep * i}`;
      grid.set(coord, (grid.get(coord) || 0) + 1);
      i += 1;
    }
  });

  return Array.from(grid.values()).filter((x) => x > 1).length;
};

File: 2025/src/days/day5.rs
use std::fs;

#[derive(Debug)]
struct Range {
    min: u64,
    max: u64,
}

pub fn run() {
    let input = fs::read_to_string("../files/2025_05_input.txt").unwrap();
    let mut paragraphs = input.split("\n\n");

    let ranges: Vec<Range> = paragraphs
        .next()
        .unwrap()
        .lines()
        .filter(|l| l.trim().len() > 0)
        .map(|l| {
            let mut vs = l.trim().split('-');

            return Range {
                min: vs.next().unwrap().parse().unwrap(),
                max: vs.next().unwrap().parse().unwrap(),
            };
        })
        .collect();

    let ingredients: Vec<u64> = paragraphs
        .next()
        .unwrap()
        .lines()
        .filter(|l| l.trim().len() > 0)
        .map(|l| l.trim().parse().unwrap())
        .collect();

    println!("{}", part1(&ranges, &ingredients));
    println!("{}", part2(&ranges));
}

fn part1(ranges: &Vec<Range>, ingredients: &Vec<u64>) -> u64 {
    let mut count = 0;

    for &ingredient in ingredients {
        if ranges
            .iter()
            .any(|r| ingredient >= r.min && ingredient <= r.max)
        {
            count += 1;
        }
    }

    return count;
}

fn part2(ranges: &Vec<Range>) -> u64 {
    let mut count = 0;
    let mut ingredient = 0;

    loop {
        let next_max = ranges
            .iter()
            .filter(|&r| ingredient >= r.min && ingredient <= r.max)
            .map(|r| r.max)
            .max();

        match next_max {
            Some(m) => {
                count += m - ingredient + 1;
                ingredient = m + 1;
                continue;
            }
            None => (),
        }

        let next_min = ranges
            .iter()
            .map(|r| r.min)
            .filter(|m| m >= &ingredient)
            .min();

        match next_min {
            Some(m) => ingredient = m,
            None => break,
        }
    }

    return count;
}
